File: crypto.cs
using System;
using System.Collections.Generic;

namespace csharp_lib_crypto
{
    using PublicKey = System.String;
    using PrivateKey = System.String;
    using Seed = System.String;
    using Address = System.String;
    using ChainId = System.Int32;

    public interface IKeyPair
    {
        PublicKey PublicKey { get; }
        PrivateKey PrivateKey { get; }
    }

    static class WavesCryptoConstants
    {
        public const int PUBLIC_KEY_LENGTH = 32;
        public const int PRIVATE_KEY_LENGTH = 32;
        public const int SIGNATURE_LENGTH = 64;
    }

    public enum WavesChainId
    {
        MAIN_NET_CHAIN_ID = 87,
        TEST_NET_CHAIN_ID = 84,
    }

    public interface IWavesCrypto
    {

        byte[] Blake2b(byte[] input);
        byte[] Keccak(byte[] input);
        byte[] Sha256(byte[] input);

        string Base58Encode(byte[] input);
        byte[] Base58Decode(string input);
        string Base64Encode(byte[] input);
        byte[] Base64Decode(string input);
        string Base16Encode(byte[] input);
        byte[] Base16Decode(string input);

        IKeyPair KeyPair(Seed seed);
        PublicKey PublicKey(Seed seed);
        PrivateKey PrivateKey(Seed seed);

        Address AddressFromPublicKey(PublicKey publicKey, ChainId? chainId = null);
        Address Address(Seed seed, ChainId? chainId = null);

        Seed RandomSeed();
        byte[] RandomBytes(int size);
        byte[] StringToBytes(string input);
        string BytesToString(byte[] input);

        byte[] SignBytesWithPrivateKey(byte[] bytes, PrivateKey privateKey);
        byte[] SignBytes(byte[] bytes, Seed seed);

        bool VerifySignature(PublicKey publicKey, byte[] bytes, byte[] signature);
        bool VerifyPublicKey(PublicKey publicKey);
        bool VerifyAddress(Address address, ChainId? chainId, PublicKey publicKey);

        byte[] SharedKey(byte[] privateKeyFrom, byte[] publicKeyTo, string prefix);
        string MessageDecrypt(byte[] sharedKey, byte[] encryptedMessage, string prefix);
        byte[] MessageEncrypt(byte[] sharedKey, string message, string prefix);
    }
}

File: crypto.swift
public typealias Bytes = [UInt8]
public typealias PublicKey = String
public typealias PrivateKey = String
public typealias Seed = String
public typealias Address = String

public protocol KeyPair {
  var publicKey: PublicKey { get }
  var privateKey: PrivateKey { get }
}

public enum WavesCryptoConstants {
  public static let publicKeyLength: Int = 32
  public static let privateKeyLength: Int = 32
  public static let signatureLength: Int = 64
}

public enum WavesChainId {
  public static let mainNet: Int = 87
  public static let testNet: Int = 84
}

public protocol WavesCrypto {
  func blake2b(input: Bytes) -> Bytes
  func keccak(input: Bytes) -> Bytes
  func sha256(input: Bytes) -> Bytes

  func base58encode(input: Bytes) -> String
  func base58decode(input: String) -> Bytes
  func base64encode(input: Bytes) -> String
  func base64decode(input: String) -> Bytes

  func keyPair(seed: Seed) -> KeyPair
  func publicKey(seed: Seed) -> PublicKey
  func privateKey(seed: Seed) -> PrivateKey

  func address(publicKey: PublicKey, chainId: String?) -> Address
  func address(seed: Seed, chainId: String?) -> Address

  func randomSeed() -> Seed

  func signBytes(bytes: Bytes, privateKey: PrivateKey) -> Bytes
  func signBytes(bytes: Bytes, seed: Seed) -> Bytes

  func verifySignature(publicKey: PublicKey, bytes: Bytes, signature: Bytes) -> Bool
  func verifyPublicKey(publicKey: PublicKey) -> Bool
  func verifyAddress(address: Address, chainId: String?, publicKey: PublicKey?) -> Bool
}
